hydra-test-utils/test/Test/HSpec/MarkdownFormatterSpec.hs

```diff
@@ -0,0 +1,95 @@
+module Test.HSpec.MarkdownFormatterSpec where
+
+import Data.List (isInfixOf)
+import Hydra.Prelude
+import System.Directory (createDirectoryIfMissing)
+import System.FilePath (splitFileName, (</>))
+import Test.Hspec.Core.Format (Event (..), Format, FormatConfig)
+import Test.Hspec.Core.Runner (
+  Config (..),
+  Summary (..),
+  defaultConfig,
+  hspecWith,
+  hspecWithResult,
+ )
+import Test.Hspec.MarkdownFormatter
+import Test.Hydra.Prelude
+import Test.QuickCheck (Positive (..), Small (..), counterexample, forAll, frequency, property, vectorOf)
+import Test.QuickCheck.Monadic (assert, forAllM, monadic, monadicIO, monitor, run)
+
+spec :: Spec
+spec =
+  around (withTempDir "foo") $ do
+    it "format markdown content as expected" $ \tmpDir -> do
+      hspecWithMarkdown tmpDir testSpec
+        `shouldReturn` "# Some Spec\
+                       \\n\n\
+                       \## Sub Spec\
+                       \\n\n\
+                       \* does one thing\n\
+                       \* does two things"
+
+    it "generates markdown content to file when running spec" $ \tmpDir ->
+      property $
+        monadicIO $
+          forAllM (genDescribe 3) $ \aSpecTree -> do
+            content <- run $ hspecWithMarkdown tmpDir (toSpec aSpecTree)
+            monitor (counterexample content)
+            assert $ all (`isInfixOf` content) $ listLabels aSpecTree
+
+hspecWithMarkdown :: FilePath -> Spec -> IO String
+hspecWithMarkdown tmpDir aSpec = do
+  let markdownFile = tmpDir </> "result.md"
+  _summary <-
+    hspecWithResult
+      defaultConfig
+        { configIgnoreConfigFile = True -- Needed to ensure we don't mess up this run with our default config
+        , configFormat = Just (markdownFormatter markdownFile)
+        }
+      aSpec
+  readFile markdownFile
+
+listLabels :: TestTree -> [String]
+listLabels = go []
+ where
+  go acc (Describe s tts) = foldMap (go (s : acc)) tts
+  go acc (It s) = s : acc
```

Comment:
Could: use `foldMap` all the way, e.g. `foldMap go` where

```
go = \case
  Describe _ s -> [s]
  It s -> [s]
```

But this requires you to write add a `deriving Foldable` to `TestTree`

Reply:
`Foldable t` requires `t :: Type -> Type` which `TestTree` is not :(

Reply:
Perhaps I could make it a `Monoid` 🤔 ?

Reply:
Nah, not worth it. Should use an editor with haskell tooling to suggest things :)